core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultMemoryHealthChecker.java

```diff
@@ -0,0 +1,51 @@
+package com.linecorp.armeria.server.healthcheck;
+
+import java.lang.management.BufferPoolMXBean;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryPoolMXBean;
+import java.lang.management.MemoryType;
+import java.lang.management.MemoryUsage;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.google.common.base.MoreObjects;
+
+public class DefaultMemoryHealthChecker implements HealthChecker {
+
+    private final double targetHeapMemoryUtilizationRate;
+    private final double targetNonHeapMemoryUtilizationRate;
+    private final double targetTotalMemoryUtilizationRate;
+
+    public DefaultMemoryHealthChecker(final double targetHeapMemoryUtilizationRate, final double targetNonHeapMemoryUtilizationRate, final double targetMemoryTotalUsage) {
+        this.targetHeapMemoryUtilizationRate = targetHeapMemoryUtilizationRate;
+        this.targetNonHeapMemoryUtilizationRate = targetNonHeapMemoryUtilizationRate;
+        this.targetTotalMemoryUtilizationRate = targetMemoryTotalUsage;
+    }
+
+    @Override
+    public boolean isHealthy() {
+        final List<MemoryPoolMXBean> heapMemories = getHeapMemories();
+        final double heapMemoryUsage = heapMemories.stream().map(MemoryPoolMXBean::getUsage).mapToDouble(MemoryUsage::getUsed).sum();
+        final double maximumHeapMemory = heapMemories.stream().map(MemoryPoolMXBean::getUsage).mapToDouble(MemoryUsage::getMax).sum();
+        final long runtimeMaxMemory = Runtime.getRuntime().maxMemory();
+
+        final BufferPoolMXBean nonHeapMemoryUsage = ManagementFactory.getPlatformMXBean(BufferPoolMXBean.class);
+        final double totalMemoryUsage = Runtime.getRuntime().totalMemory();
+        return (heapMemoryUsage / maximumHeapMemory) <= targetHeapMemoryUtilizationRate &&
+               (nonHeapMemoryUsage == null || (nonHeapMemoryUsage.getMemoryUsed() / nonHeapMemoryUsage.getTotalCapacity()) <= targetNonHeapMemoryUtilizationRate) &&
+               (runtimeMaxMemory == Long.MAX_VALUE || totalMemoryUsage / runtimeMaxMemory <= targetTotalMemoryUtilizationRate);
```

Comment:
Can we log a warning message if `nonHeapMemoryUsage` is null or `runtimeMaxMemory` is `Long.MAX_VALUE`, so that we tell our users that those properties are not available and thus it will never become unhealthy even if you specified the threshold?

We could keep a boolean flag so make sure that the log message is logged only once, so we don't pollute the log file.